v2/kr/lunar.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// hasLeapMonth `year` has leap month or not
+func hasLeapMonth(year int) bool {
+    return leapMonth(year) != 0
+}
+
+// isLeapMonth is leap month or not
+func isLeapMonth(year, month int) bool {
+    return leapMonth(year) == month
+}
+
+// getLunarInfo get pre-calcuated data by lunar year
+func getLunarInfo(year int) int {
+    idx := year - MINYEAR
+    if idx < 0 || idx > len(LunarInfoList) {
+        return 0
+    }
+    return LunarInfoList[idx]
+}
+
+// Solar2Lunar converts to lunar year/month/day and if it is belongs to the leap month
+// Korean lunar calendar
+func Solar2Lunar(t time.Time) (ymd [3]int, isLeap bool) {
+    const (
+        BASEYEAR = 1391
+        MINDATE  = 2229156 // 1391-02-05 ( lunisolar 1391-01-01 )
+        MAXDATE  = 2470172 // 2050-12-31 ( lunisolar 2050-11-18 )
+    )
+
+    // TODO gregorian 1582-10~05 ~ 1582-10-14 dates do not exist.
+    days := julian(t)
+    if days < MINDATE || days > MAXDATE {
+        log.Fatalf("The date is out of range: %d", days)
+    }
+
+    days -= MINDATE
+    month := sort.Search(len(MonthTable), func(i int) bool { return MonthTable[i] > days }) - 1
+    if month > len(MonthTable)-1 {
+        log.Fatalf("Out of MonthTable range: %d/%d", month, len(MonthTable))
+    }
+
+    year := sort.Search(len(YearTable), func(i int) bool { return YearTable[i] > month }) - 1
+    if year > len(YearTable)-1 {
+        log.Fatalf("Out of YearTable range: %d/%d", year, len(YearTable))
+    }
+
+    var day int
+    month, day = month-YearTable[year]+1, days-MonthTable[month]+1
+    if LeapTable[year] != 0 && LeapTable[year] <= month {
+        if LeapTable[year] == month {
+            isLeap = true
+        } else {
+            isLeap = false
+        }
+        month--
+    } else {
+        isLeap = false
+    }
+
+    ymd[0], ymd[1], ymd[2] = year+BASEYEAR, month, day
+    return ymd, isLeap
+}
+
+// Lunar2Solar converts to solar calendar time
+func Lunar2Solar(year, month, day int, isLeap bool) time.Time {
+    const (
+        BASEYEAR = 1391
+        MINDATE  = 2229156 // 1391-02-05 ( lunisolar 1391-01-01 )
+        MAXDATE  = 2470172 // 2050-12-31 ( lunisolar 2050-11-18 )
+    )
+
+    year -= BASEYEAR
+    if year < 0 || year >= len(YearTable) {
+        log.Fatal("Year is out of range")
+    }
+
+    if month < 1 || month > 12 {
+        log.Fatal("Month is out of range")
+    }
+
+    if isLeap && LeapTable[year]-1 != month {
+        log.Fatal("Wrong leap month")
+    }
+
+    months := YearTable[year] + month - 1
+    if isLeap && month+1 == LeapTable[year] {
+        months++
+    } else if LeapTable[year] != 0 && LeapTable[year] < month {
+        months++
+    }
+
+    days := MonthTable[months] + day - 1
+    if day < 1 || days >= MonthTable[months+1] {
+        log.Fatal("Wrong day")
+    }
+
+    y, m, d := julianDayToDate(days + MINDATE)
+    return time.Date(y, time.Month(m), d, 0, 0, 0, 0, cal.DefaultLoc)
+}
+
+func julianDayToDate(jd int) (year, month, day int) {
```

Comment:
There are already Julian functions in cal_funcs.go. Do those not work?

Reply:
oh i missed it.